packages/liveblocks-react/src/umbrella-store.ts

```diff
@@ -1081,36 +1105,56 @@ function internalToExternalState<M extends BaseMetadata>(
         break;
       }
       case "mark-inbox-notification-as-read": {
-        output.inboxNotifications[optimisticUpdate.inboxNotificationId] = {
-          ...state.inboxNotificationsById[optimisticUpdate.inboxNotificationId],
-          readAt: optimisticUpdate.readAt,
-        };
+        const ibn =
+          computed.inboxNotificationsById[optimisticUpdate.inboxNotificationId];
+
+        // If the inbox notification doesn't exist in the cache, we do not apply the update
+        if (ibn === undefined) {
+          break;
+        }
+
+        computed.inboxNotificationsById[optimisticUpdate.inboxNotificationId] =
+          { ...ibn, readAt: optimisticUpdate.readAt };
         break;
       }
       case "mark-all-inbox-notifications-as-read": {
-        for (const id in output.inboxNotifications) {
-          output.inboxNotifications[id] = {
-            ...output.inboxNotifications[id],
+        for (const id in computed.inboxNotificationsById) {
+          const ibn = computed.inboxNotificationsById[id];
+
+          // If the inbox notification doesn't exist in the cache, we do not apply the update
+          if (ibn === undefined) {
+            break;
+          }
+
+          computed.inboxNotificationsById[id] = {
+            ...ibn,
             readAt: optimisticUpdate.readAt,
           };
         }
         break;
       }
       case "delete-inbox-notification": {
-        const {
-          [optimisticUpdate.inboxNotificationId]: _,
-          ...inboxNotifications
-        } = output.inboxNotifications;
-        output.inboxNotifications = inboxNotifications;
+        delete computed.inboxNotificationsById[
+          optimisticUpdate.inboxNotificationId
+        ];
         break;
       }
       case "delete-all-inbox-notifications": {
-        output.inboxNotifications = {};
+        computed.inboxNotificationsById = {};
         break;
       }
+
       case "update-notification-settings": {
-        output.notificationSettings[optimisticUpdate.roomId] = {
-          ...output.notificationSettings[optimisticUpdate.roomId],
+        const settings =
+          computed.notificationSettingsByRoomId[optimisticUpdate.roomId];
+
+        // If the inbox notification doesn't exist in the cache, we do not apply the update
+        if (settings === undefined) {
+          break;
+        }
```

Comment:
Was one of the bugs caused by us not checking if an inbox notification existed before applying any optimistic update to it?

Reply:
Correct! Since we weren't catching the `foo[0]` undefined index access here, it was possible (in theory, not sure if it's an actual practical problem), that we'd spread `{ ...undefined, updatedAt: new Date() }` in there, effectively creating an incorrect resource with only an `updatedAt` field and injecting that in the store state.